op-program/host/l2_source.go

```diff
@@ -0,0 +1,139 @@
+	if err != nil {
+		return nil, err
+	}
+	experimentalL2ClientCfg := sources.L2ClientDefaultConfig(config.Rollup, true)
+	experimentalL2Client, err := NewL2Client(experimentalRPC, logger, nil, &L2ClientConfig{L2ClientConfig: experimentalL2ClientCfg, L2Head: config.L2Head})
+	if err != nil {
+		return nil, err
+	}
+
+	experimentalClient := NewL2ExperimentalClient(experimentalL2Client, experimentalRPC)
+	source.experimentalClient = experimentalClient
+	source.experimentalEnabled = true
+
+	return source, nil
+}
+
+// CodeByHash implements prefetcher.L2Source.
+func (l *L2Source) CodeByHash(ctx context.Context, hash common.Hash) ([]byte, error) {
+	if l.experimentalEnabled {
+		// This means experimental source was not able to retrieve relevant information from eth_getProof or debug_executionWitness
+		// We should fall back to the canonical source, and log a warning, and record a metric
+		l.logger.Warn("Experimental source failed to retrieve code by hash, falling back to canonical source", "hash", hash)
+	}
+	return l.canonicalDebugClient.CodeByHash(ctx, hash)
+}
+
+// NodeByHash implements prefetcher.L2Source.
+func (l *L2Source) NodeByHash(ctx context.Context, hash common.Hash) ([]byte, error) {
+	if l.experimentalEnabled {
+		// This means experimental source was not able to retrieve relevant information from eth_getProof or debug_executionWitness
+		// We should fall back to the canonical source, and log a warning, and record a metric
+		l.logger.Warn("Experimental source failed to retrieve node by hash, falling back to canonical source", "hash", hash)
+	}
+	return l.canonicalDebugClient.NodeByHash(ctx, hash)
+}
+
+// InfoAndTxsByHash implements prefetcher.L2Source.
+func (l *L2Source) InfoAndTxsByHash(ctx context.Context, blockHash common.Hash) (eth.BlockInfo, types.Transactions, error) {
+	if l.experimentalEnabled {
+		return l.experimentalClient.InfoAndTxsByHash(ctx, blockHash)
+	}
+	return l.canonicalEthClient.InfoAndTxsByHash(ctx, blockHash)
+}
+
+// OutputByRoot implements prefetcher.L2Source.
+func (l *L2Source) OutputByRoot(ctx context.Context, root common.Hash) (eth.Output, error) {
+	if l.experimentalEnabled {
+		return l.experimentalClient.OutputByRoot(ctx, root)
+	}
+	return l.canonicalEthClient.OutputByRoot(ctx, root)
+}
+
+// ExecutionWitness implements prefetcher.L2Source.
+func (l *L2Source) ExecutionWitness(ctx context.Context, blockNum uint64) (*eth.ExecutionWitness, error) {
+	if !l.experimentalEnabled {
+		l.logger.Error("Experimental source is not enabled, cannot fetch execution witness", "blockNum", blockNum)
+		panic("experimental source is not enabled")
+	}
+	return l.experimentalClient.ExecutionWitness(ctx, blockNum)
+}
+
+// GetProof implements prefetcher.L2Source.
+func (l *L2Source) GetProof(ctx context.Context, address common.Address, storage []common.Hash, blockTag string) (*eth.AccountResult, error) {
+	if l.experimentalEnabled {
+		l.experimentalClient.GetProof(ctx, address, storage, blockTag)
```

Comment:
did you mean to return here?

Reply:
good catch, this method wasn't working on my client so I had removed it temporarily. Fixed now!